pkg/controllers/customcluster_controller.go

```diff
@@ -47,16 +77,392 @@ type CustomClusterController struct {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the customCluster instance.
+	customCluster := &v1alpha1.CustomCluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, customCluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{Requeue: false}, nil
+		}
+		return ctrl.Result{RequeueAfter: RequeueAfter}, err
+	}
+
+	// Fetch the CustomMachine instance.
+	customMachinekey := client.ObjectKey{
+		Namespace: customCluster.Spec.MachineRef.Namespace,
+		Name:      customCluster.Spec.MachineRef.Name,
+	}
+	customMachine := &v1alpha1.CustomMachine{}
+	if err := r.Client.Get(ctx, customMachinekey, customMachine); err != nil {
+		log.Error(err, "can not get customMachine")
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{Requeue: false}, nil
+		}
+		return ctrl.Result{RequeueAfter: RequeueAfter}, err
+	}
+
+	// Fetch the Cluster instance
+	clusterkey := client.ObjectKey{
+		Namespace: customCluster.Spec.ClusterRef.Namespace,
+		Name:      customCluster.Spec.ClusterRef.Name,
+	}
+	cluster := &clusterv1.Cluster{}
+	if err := r.Client.Get(ctx, clusterkey, cluster); err != nil {
+		log.Error(err, "can not get cluster")
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{Requeue: false}, nil
+		}
+		return ctrl.Result{RequeueAfter: RequeueAfter}, err
+	}
+
+	// Fetch the KubeadmControlPlane instance.
+	kcpKey := client.ObjectKey{
+		Namespace: cluster.Spec.ControlPlaneRef.Namespace,
+		Name:      cluster.Spec.ControlPlaneRef.Name,
+	}
+	kcp := &controlplanev1.KubeadmControlPlane{}
+	if err := r.Client.Get(ctx, kcpKey, kcp); err != nil {
+		log.Error(err, "can not get kcp")
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{Requeue: false}, nil
+		}
+		return ctrl.Result{RequeueAfter: RequeueAfter}, err
+	}
+
+	// TODO: check cluster status
+
+	log = log.WithValues("customCluster", klog.KObj(customCluster))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	return r.reconcile(ctx, customCluster, customMachine, cluster, kcp)
+}
+
+func (r *CustomClusterController) reconcile(ctx context.Context, customCluster *v1alpha1.CustomCluster, customMachine *v1alpha1.CustomMachine, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Delete current cm and create new cm using current object
+	if err := r.UpdateHostsConfigMap(ctx, customCluster, customMachine); err != nil {
+		log.Error(err, "failed to create hosts configMap")
+		return ctrl.Result{RequeueAfter: RequeueAfter}, err
+	}
+
+	if err := r.UpdateVarsConfigMap(ctx, customCluster, customCluster, cluster, kcp); err != nil {
+		log.Error(err, "failed to create vars configMap")
+		return ctrl.Result{RequeueAfter: RequeueAfter}, err
+	}
+
+	initClusterJob := r.CreateKubesprayInitClusterJob(customCluster)
 
-	// TODO(user): your logic here
+	if curJob, err := r.ClientSet.BatchV1().Jobs(initClusterJob.Namespace).Create(context.Background(), initClusterJob, metav1.CreateOptions{}); err != nil {
```

Comment:
Should consider cluster status to prevent create the job again, but here we just return an error